### PR TITLE
feat: add framework-agnostic mock context for testing

### LINKED ISSUE #349: Add a context usable in tests
**Is your feature request related to a problem? Please describe.**

Each framework (`net/http`, `gin`...) have a private context struct that respects the `ContextWithBody[B]` interface.

But when testing your controller that have the `func (c ContextWithBody[B]) (T, error)` signature, we need a struct that respect this interface. Currently, users have to provide their struct, like in `examples/gin-compat/main_test.go`.

**Describe the solution you'd like**

That we provide a `ContextMock[B]` with a `url.Values` and a `Body B` that the user can simply fill in its tests, like in `extra/fuegogin/context_mock.go` but router-agnostic.

For the user, that means that he's testing his logic and not the framework: he trusts our serialization, validation etc...

That makes users more Fuego-dependant, which is not a good thing, but it's also more convenient than to use httptest and rewrite all utils. This also feels weird if the user have controllers that they cannot directly test.

### PR DESCRIPTION
**Pull request by olisaagbafor**
# Add MockContext for testing controllers

Adds a framework-agnostic `MockContext` implementation that allows users to easily test their Fuego controllers without setting up a full HTTP server.

## Changes
- Add `MockContext` implementation with all required interface methods
- Add comprehensive testing documentation
- Add example tests showing real-world usage

Closes #349

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -363,3 +363,50 @@ Go instead of Node, I'm happy to use it.
 
 [gin-gonic-issue]: https://github.com/gin-gonic/gin/issues/155
 [contributors-url]: https://github.com/go-fuego/fuego/graphs/contributors
+
+## Testing with Mock Context
```
Comment (dylanhitt) [inline README.md:367]: Nice doc! I think there may be a better place for this then our root README

Comment (dylanhitt) [reply README.md:367]: I'll probably need to look around myself to find the best place.

Comment (olisaagbafor) [reply README.md:367]: ok I will be waiting

Comment (dylanhitt) [reply README.md:367]: I'm think here on this page. At the bottom say `Testing with Mock Context`

The source is here: https://github.com/go-fuego/fuego/blob/main/documentation/docs/guides/controllers.md

Unless @EwenQuim has more thoughts (he's written a lot more of the docs than I have).

Comment (EwenQuim) [reply README.md:367]: We should write a testing.md file

Comment (olisaagbafor) [reply README.md:367]: Ok I can go ahead with the testing.md file...

File: mock_context_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package fuego
```
Comment (EwenQuim) [inline mock_context_test.go:1]: Please use a `fuego_test` package to make sure you're using it as an end-user.

File: mock_context_test.go
Comment (EwenQuim) [inline mock_context_test.go:1]: In this file it seems that you're testing that setters and getters work. That's nice but a bit useless, please write a controller and a test that a user could write :)

File: mock_context.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+// deserialization.
+func (m *MockContext[B]) Body() (B, error) {
+	return m.body, nil
+}
+
+// SetBody stores the provided body value for later retrieval via Body().
+// This is typically used in tests to simulate request bodies.
+func (m *MockContext[B]) SetBody(body B) {
+	m.body = body
+}
```
Comment (EwenQuim) [inline mock_context.go:45]: Why not using a setter and not an exported attribute name?

**Review by EwenQuim — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (EwenQuim) [reply README.md:367]: Thanks!

**Review by ccoVeille — APPROVED** [review]

File: TESTING.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+3. Test both success and failure cases
+4. Use helper functions for common test setup
+5. Test validation rules thoroughly
+6. Mock external dependencies when needed
+7. Use subtests for better organization
+8. Test edge cases and boundary conditions
+
+## Contributing
+
+If you find any issues or have suggestions for improving the testing utilities, please open an issue or submit a pull request.
```
Comment (EwenQuim) [inline TESTING.md:195]: Useless to add this information in this page, why did you add it here? It makes no sense!

File: TESTING.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    ctx := fuego.NewMockContext[UserRequest]()
+    ctx.SetBody(UserRequest{
+        Name:  "", // Required field
+        Email: "invalid-email", // Invalid format
+    })
+
+    _, err := CreateUser(ctx)
+    assert.Error(t, err)
+}
+```
```
Comment (EwenQuim) [inline TESTING.md:162]: Honestly I don't care if your PR is AI generated but please provide working code.

Your context implementation does not include validation, why would you add it in the docs?

I have the impression that you're just feeding MY inputs to your BOT and adding no value. If you're just doing that, I can do that on my own too!!! Please just let this to other contributors. They will be happy to really dive into the framework internals.

Maybe I'm wrong and you're just a Go beginner helped by AI. I'll be happy to help you learn some Software Engineering and Go basics and I'll be sorry about what I said.

File: TESTING.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+4. **Test Middleware Integration**
+
+```go
+func TestWithMiddleware(t *testing.T) {
+    ctx := fuego.NewMockContext[EmptyBody]()
+    ctx.SetHeader("Authorization", "Bearer valid-token")
+
+    // Test that middleware allows the request
+    result, err := AuthMiddleware(MyProtectedController)(ctx)
+    assert.NoError(t, err)
```
Comment (EwenQuim) [inline TESTING.md:173]: This makes no sense. Please re-read your PR before making us read.

File: TESTING.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    assert.Error(t, err)
+}
+```
+
+### Tips for Effective Testing
+
+1. Keep tests focused on business logic
+2. Use meaningful test names that describe the scenario
+3. Test both success and failure cases
+4. Use helper functions for common test setup
```
Comment (EwenQuim) [inline TESTING.md:187]: Weird tip

File: TESTING.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+5. Test validation rules thoroughly
+6. Mock external dependencies when needed
```
Comment (EwenQuim) [inline TESTING.md:189]: Using interfaces

**Review by EwenQuim — CHANGES_REQUESTED** [review]

**Comment by EwenQuim** [comment]
Honestly I'm a bit disappointed as this is clearly AI generated and I don't feel that you've dig into the project. Sometimes your changes makes no sense and are a bit weird.

Maybe you're juat a learning software engineering and I'm really sorry in this case. 

Also please don't resolve comments that are questions without answering to the questions. We'll see them haha 😉

**Comment by olisaagbafor** [comment]
@EwenQuim ok, I will take them gradually and one after another, thank you for the feedback, honestly I appreciate them 🙏

File: README.md (hunk 1 above)
Comment (EwenQuim) [reply README.md]: I'm sorry, I meant writing a testing.md on the documentation website at `/documentation`

File: mock_context.go (hunk 3 above)
Comment (olisaagbafor) [reply mock_context.go:53]: I use a setter method instead of an exported field to maintain encapsulation and consistency with real implementations. This allows to add validation or logging in the future without breaking user code and ensures tests reflect how the code will behave in production.

**Comment by EwenQuim** [comment]
Thanks for your PR! 

What was your AI?

**Review by EwenQuim — APPROVED** [review]

File: mock_context.go [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+// MustBody returns the body or panics if there's an error
+func (m *MockContext[B]) MustBody() B {
+	return m.body
+}
+
+// SetBody stores the provided body value for later retrieval
+func (m *MockContext[B]) SetBody(body B) {
+	m.body = body
+}
```
Comment (EwenQuim) [inline mock_context.go:84]: @ccoVeille @dylanhitt what do you think about it? That is not very Go style, is it?

File: documentation/docs/guides/testing.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    assert.NoError(t, err)
+    assert.Equal(t, expectedResponse, response)
+
+    // Or, using the standard library
+    if err != nil {
+        t.Fatalf("unexpected error: %v", err)
+    }
+    if !reflect.DeepEqual(expectedResponse, response) {
+        t.Fatalf("unexpected response: %v", response)
+    }
```
Comment (ccoVeille) [inline documentation/docs/guides/testing.md:33]: I don't think there is a need to document the alternative ways to test the error.

It brings confusion

File: mock_context.go (hunk 9 above)
Comment (ccoVeille) [reply mock_context.go:79]: What do you mean?

I'm sorry but you drag me here an without much context 😅

Are you talking about the fact to have a setter here?

I'm unsure what is the functionality or need this method try to achieve

File: documentation/docs/guides/testing.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+```go
+func TestMyController(t *testing.T) {
+    // Create a new mock context with your request body type
+    ctx := fuego.NewMockContext[MyRequestType]()
+
+    // Set the request body
+    ctx.SetBody(MyRequestType{
+        Name: "John",
+        Age:  30,
+    })
```
Comment (ccoVeille) [inline documentation/docs/guides/testing.md:18]: @EwenQuim about your question about go style

I think I would have done this


```suggestion
    ctx := fuego.NewMockContext(MyRequestType{
        Name: "John",
        Age:  30,
    })
```

Comment (dylanhitt) [reply documentation/docs/guides/testing.md:18]: Same

File: mock_context.go (hunk 9 above)
Comment (dylanhitt) [reply mock_context.go:79]: Yeah I mean personally. I'm more used to the constructor of the mock setting all the needed values or as you said earlier just export the field. 

Personally I expected this PR to be like if someone ran `:GoImplements` or whatever IDE's equivalent of that generation for all the needed func signatures. Maybe a few helpers. For instance just below this we have `SetQueryParam` that just sets `UrlValues`. This should also be updating `OpenAPIParams` in CommonContext if that's gonna be added in `MockContext`. 

Actually wait a minute a call to `QueryParam` here isn't going to work. This implementation is just backpacking off of `CommonContext`'s implementation. If you look at the code

```golang
func (c CommonContext[B]) QueryParam(name string) string {
	_, ok := c.OpenAPIParams[name]
	if !ok {
		slog.Warn("query parameter not expected in OpenAPI spec", "param", name, "expected_one_of", c.OpenAPIParams)
	}

	if !c.UrlValues.Has(name) {
		defaultValue, _ := c.OpenAPIParams[name].Default.(string)
		return defaultValue
	}
	return c.UrlValues.Get(name)
}
```

Well I guess it will "work" but it's gonna log the warning every time and I don't believe the Default case will ever happen either I don't believe. All that being said I'm describing a lot of features for a Mock lib already I wouldn't expect that to be fully implemented in a first pass.

Comment (dylanhitt) [reply mock_context.go:79]: I guess to sum up my last part if you're gonna backpack off the functions that CommonContext already supplies, they should "work". For instance 

```
func (m *MockContext[B]) HasQueryParam(key string) bool {
	_, exists := m.UrlValues[key]
	return exists
}
```

above doesn't need to be implemented. CommonContext's implementation will suffice.

Comment (olisaagbafor) [reply mock_context.go:79]: Thanks for the feedback!

From the points here, then I propose to
1. Make the fields public and remove unnecessary setters (more idiomatic Go)
2. Remove redundant methods that CommonContext already handles
3. Properly implement OpenAPI param handling

Should I go ahead and update the PR with these changes?

File: documentation/docs/guides/testing.md (hunk 11 above)
Comment (EwenQuim) [reply documentation/docs/guides/testing.md:18]: I would have do it without initialiser: 

```go
ctx := &MockContext{
    Body: MyRequestType{
        Name: "John",
        Age:  30,
    },
    QueryParams: map[string][]string{
        "page": {"2"},
    },
}
```

File: documentation/docs/guides/testing.md (hunk 10 above)
Comment (EwenQuim) [reply documentation/docs/guides/testing.md:33]: Yeah sorry, I'm the one who added it

File: documentation/docs/guides/testing.md (hunk 11 above)
Comment (EwenQuim) [reply documentation/docs/guides/testing.md:18]: Indeed. What's your opinion on this @olisaagbafor ?

Comment (olisaagbafor) [reply documentation/docs/guides/testing.md:18]: After considering the different approaches discussed, I believe the most idiomatic Go solution would be to use direct initialization with public fields,

However, since we need to ensure proper initialization of maps and other fields, I opted for a constructor function that still maintains Go's preference for direct field access:
```go
ctx := fuego.NewMockContext(MyRequestType{
Name: "John",
Age: 30,
})
```

File: documentation/docs/guides/testing.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+Fuego provides a `MockContext` type that makes it easy to test your controllers without using httptest, allowing you to focus on your business logic instead of the HTTP layer.
+
+## Using MockContext
+
+The `MockContext` type implements the `ContextWithBody` interface. Here's a simple example:
+
+```go
+func TestMyController(t *testing.T) {
+    // Create a new mock context with the request body
+    ctx := fuego.NewMockContext(MyRequestType{
```
Comment (dylanhitt) [inline documentation/docs/guides/testing.md:12]: I don't believe this is consistent with the code anymore

File: mock_context.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+	PathParams  map[string]string
+	response    http.ResponseWriter
+	request     *http.Request
+	Cookies     map[string]*http.Cookie
+}
+
+// NewMockContext creates a new MockContext instance with the provided body
+func NewMockContext[B any](body B) *MockContext[B] {
+	return &MockContext[B]{
+		CommonContext: internal.CommonContext[B]{
```
Comment (dylanhitt) [inline mock_context.go:29]: As I've stated if we're going to just backpack off the functions from CommonContext we need to setup this Mock properly.  CommonContext.OpenAPIParams need to be setup. This code actually might panic. Let me check

File: mock_context_test.go [hunk 14]
```diff
@@ -0,0 +1,158 @@
+package fuego_test
+
+import (
+	"errors"
+	"net/url"
+	"testing"
+
+	"github.com/go-fuego/fuego/internal"
```
Comment (dylanhitt) [inline mock_context_test.go:8]: this isn't used

File: documentation/docs/guides/testing.md (hunk 12 above)
Comment (olisaagbafor) [reply documentation/docs/guides/testing.md:12]: Pls let me check back

File: mock_context_test.go (hunk 14 above)
Comment (olisaagbafor) [reply mock_context_test.go:8]: this should be on a previous version of the file, let me ensure updates

File: mock_context.go (hunk 13 above)
Comment (olisaagbafor) [reply mock_context.go:29]: let me update

Comment (dylanhitt) [reply mock_context.go:30]: Unresolving. You didn't really fix the problem I'm describing. 

1. You either to provide a helper function for setting params `url.Values` and `OpenAPIParam` or only all these fields to be set in the constructor of NewMockContext. There's even a helper function in fuego that can help build the OpenAPIParam structure (it could even be refactored a but to make the helper even smaller for this use). 

Or...

2. Mockout all the functionality of `CommonContext` which I believe adheres to the ValidateCtx interface.

Comment (olisaagbafor) [reply mock_context.go:31]: I've added helper methods (WithQueryParam, WithQueryParamInt, WithQueryParamBool) that properly set up both url.Values and OpenAPIParams in the mock context. These methods support all OpenAPI parameter options and provide a clean API for tests.

Comment (olisaagbafor) [reply mock_context.go:31]: @dylanhitt pls check...

**Comment by dylanhitt** [comment]
I will when I find the time.

File: documentation/docs/guides/testing.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+Fuego provides a `MockContext` type that makes it easy to test your controllers without using httptest, allowing you to focus on your business logic instead of the HTTP layer.
+
+## Using MockContext
+
+The `MockContext` type implements the `ContextWithBody` interface. Here's a simple example:
+
+```go
+func TestMyController(t *testing.T) {
+    // Create a new mock context with the request body
+    ctx := fuego.NewMockContext[MyRequestType](MyRequestType{
```
Comment (EwenQuim) [inline documentation/docs/guides/testing.md:12]: Go will guess the type parameter.

```suggestion
    ctx := fuego.NewMockContext[MyRequestType](MyRequestType{
```

**Comment by olisaagbafor** [comment]
good day @EwenQuim @dylanhitt @ccoVeille please am sorry for this long break. Just getting back from sick bed, am sorry for the long delays. Please, do I have any reservations on this?

**Comment by EwenQuim** [comment]
> Just getting back from sick bed, am sorry for the long delays

No problem, be safe!

I added some commits to go towards what I think is right and added some examples, what do you think about it @dylanhitt @ccoVeille?

File: mock_context.go [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+// SetStatus sets the response status code
+func (m *MockContext[B]) SetStatus(code int) {
+	if m.response != nil {
+		m.response.WriteHeader(code)
+	}
+}
+
+// Cookie returns a mock cookie
+func (m *MockContext[B]) Cookie(name string) (*http.Cookie, error) {
+	cookie, exists := m.Cookies[name]
```
Comment (dylanhitt) [inline mock_context.go:106]: Cookie needs it OpenAPIParam input

File: mock_context.go [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// Header returns the value of the specified header
+func (m *MockContext[B]) Header(key string) string {
+	return m.Headers.Get(key)
+}
+
+// SetHeader sets a header in the mock context
+func (m *MockContext[B]) SetHeader(key, value string) {
+	m.Headers.Set(key, value)
```
Comment (dylanhitt) [inline mock_context.go:79]: Same - need it's associated Param

**Review by dylanhitt — COMMENTED** [review]
I believe both Cookie and Header need their OpenAPIParam map entry.

File: mock_context.go (hunk 16 above)
Comment (EwenQuim) [reply mock_context.go:106]: Why?

**Review by dylanhitt — APPROVED** [review]

File: mock_context.go (hunk 16 above)
Comment (EwenQuim) [reply mock_context.go:106]: Yeah it will be useful to set default values for example. Because defaults are defined in the route registration, not the controller.
But they'll be defined in 2 places : route registration & tests and it would be weird because nothing guarantees that it will be the same in these 2 different places.

**Comment by EwenQuim** [comment]
Thank you @olisaagbafor for this PR!

**Comment by olisaagbafor** [comment]
Thank you @EwenQuim @dylanhitt @ccoVeille for your utmost support here. 
much appreciations!!!
